src/metric_engine/src/index/cache.rs

```diff
@@ -0,0 +1,1293 @@
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::{
+    collections::{HashMap, HashSet},
+    path::Path,
+    sync::Arc,
+    time::{Duration, SystemTime},
+};
+
+use anyhow::Context;
+use arrow::{
+    array::{
+        Array, ArrayRef, BinaryArray, BinaryBuilder, ListArray, UInt64Array, UInt64Builder,
+        UInt8Array, UInt8Builder,
+    },
+    buffer::OffsetBuffer,
+    datatypes::{DataType, Field, Schema, ToByteSlice},
+    record_batch::RecordBatch,
+};
+use dashmap::DashMap;
+use futures::StreamExt;
+use horaedb_storage::{
+    config::StorageConfig,
+    storage::{
+        CloudObjectStorage, ScanRequest, StorageRuntimes, TimeMergeStorageRef, WriteRequest,
+    },
+    types::{ObjectStoreRef, TimeRange, Timestamp},
+};
+use tokio::{
+    sync::{
+        mpsc::{self, Receiver, Sender},
+        RwLock,
+    },
+    time::timeout,
+};
+use tracing::{error, warn};
+
+use crate::types::{
+    hash, FieldName, FieldType, Label, MetricId, MetricName, Result, SegmentDuration, SeriesId,
+    SeriesKey, TagName, TagNames, TagValue, TagValues, DEFAULT_FIELD_NAME, DEFAULT_FIELD_TYPE,
+};
+
+const COLUMN_DURATION: &str = "duration";
+const COLUMN_METRIC_NAME: &str = "metric_name";
+const COLUMN_METRIC_ID: &str = "metric_id";
+const COLUMN_SERIES_ID: &str = "series_id";
+const COLUMN_FIELD_ID: &str = "field_id";
+const COLUMN_FIELD_NAME: &str = "field_name";
+const COLUMN_FIELD_TYPE: &str = "field_type";
+const COLUMN_TAG_NAMES: &str = "tag_names";
+const COLUMN_TAG_VALUES: &str = "tag_values";
+const COLUMN_TAG_NAME: &str = "tag_name";
+const COLUMN_TAG_VALUE: &str = "tag_value";
+const COLUMN_TAG_ITEM: &str = "item";
+
+type ConcurrentMetricMap = RwLock<HashMap<MetricName, (FieldName, FieldType)>>;
+type ConcurrentSeriesMap = RwLock<HashMap<SeriesId, SeriesKey>>;
+type ConcurrentTagKVMap =
+    RwLock<HashMap<TagName, HashMap<TagValue, HashMap<MetricId, HashSet<SeriesId>>>>>;
+
+struct MetricsCache {
+    cache: DashMap<SegmentDuration, ConcurrentMetricMap>,
+    pub storage: TimeMergeStorageRef,
+    sender: Sender<Task>,
+}
+struct SeriesCache {
+    cache: DashMap<SegmentDuration, ConcurrentSeriesMap>,
+    pub storage: TimeMergeStorageRef,
+    sender: Sender<Task>,
+}
+
+#[derive(PartialEq, Eq, Hash, Debug)]
+struct SegmentSeries {
+    segment: SegmentDuration,
+    series_id: SeriesId,
+}
+
+struct TagIndexCache {
+    cache: DashMap<SegmentDuration, ConcurrentTagKVMap>,
```

Comment:
Two thoughts here:
1. We should prefer std map over dash map, those third party may introduce unexpected bugs.
2. For current segment cache, we can use an independent field to represent, so we can save one hashmap lookup.

Other questions we need to consider: 
1. How will you evict segment?